TypeCobol/Compiler/Nodes/Data.cs

```diff
@@ -282,31 +283,193 @@ public virtual DataType PrimitiveDataType
                 {
-                    _length = ((DataDefinitionEntry) this.CodeElement).Length;
+                    return _slackBytes.Value;
+                }
+
+                int index = Parent.ChildIndex(this);
+                long occupiedMemory = 0;
+                DataDefinition parent = Parent as DataDefinition;
+                _slackBytes = 0;
+
+                if (IsSynchronized && Usage != null && Usage != DataUsage.None && parent != null)
+                {
+                    while (parent != null && parent.Type != CodeElementType.SectionHeader)
+                    {
+
+                        for (int i = 0; i < index; i++)
+                        {
+                            occupiedMemory += ((DataDefinition)parent.Children[i]).PhysicalLength;
+                            occupiedMemory += ((DataDefinition)parent.Children[i]).SlackBytes;
+                        }
+
+                        index = parent.Parent.ChildIndex(parent);
+                        parent = parent.Parent as DataDefinition;
+                    }
+
+                    
+                    int m = 1;
+
+                    switch (Usage.Value)
+                    {
+                        case DataUsage.Binary:
+                            Types.PictureValidator pv = new Types.PictureValidator(this.Picture.Value, SignIsSeparate);
+                            pv.IsValid();
+                            if (pv.ValidationContext.Digits <= 4)
+                            {
+                                m = 2;
+                            }
+                            else
+                            {
+                                m = 4;
+                            }
+                            break;
+                        case DataUsage.Index:
+                        case DataUsage.Pointer:
+                        case DataUsage.ProcedurePointer:
+                        case DataUsage.ObjectReference:
+                        case DataUsage.FunctionPointer:
+                        case DataUsage.PackedDecimal:
+                            m = 4; 
+                            break;
+                        case DataUsage.FloatingPoint:
+                            m = 8;
+                            break;
+                    }
+
+                    if (occupiedMemory % m > 0)
+                    {
+                        _slackBytes = m - (occupiedMemory % m);
+                    }
+
+
+                }
+
+                return _slackBytes.Value;
+
+            }
+        }
+
+        private long? _startPosition = null;
+        public virtual long StartPosition
+        {
+            get
+            {
+                if (_startPosition.HasValue)
+                {
+                    return _startPosition.Value;
+                }
+
+                if (this is DataRedefines)
+                {
+                    SymbolReference redefined = (CodeElement as DataRedefinesEntry)?.RedefinesDataName;
```

Comment:
You can use a direct cast here, because you have checked the type of dataDefinition  with `if (dataDefinition is DataRedefines)`